File: engine/runner/src/main/scala/org/enso/runner/CurrentVersion.scala
package org.enso.runner

import org.enso.semver.SemVer

/** A helper object that allows to access current version of the runner.
  *
  * The current version is parsed from [[buildinfo]], but in development mode it
  * can be overridden by setting `enso.version.override` property. This is used
  * in project-manager tests to override the version of projects created using
  * the runner.
  */
object CurrentVersion {

  /** The version that the application should report. */
  lazy val version: SemVer = computeVersion()

  private def computeVersion(): SemVer = {
    val buildVersion =
      SemVer.parse(buildinfo.Info.ensoVersion).getOrElse {
        throw new IllegalStateException(
          "Fatal error: Enso version included in buildinfo is not a valid " +
          "semver string, this should never happen."
        )
      }
    if (buildinfo.Info.isRelease) buildVersion
    else
      sys.props
        .get("enso.version.override")
        .flatMap(v => SemVer.parse(v).toOption)
        .getOrElse(buildVersion)
  }
}

File: engine/runner/src/main/java/org/enso/runner/CurrentVersion.java
package org.enso.runner;

import org.enso.semver.SemVer;
import org.enso.version.BuildVersion;

/**
 * A helper class that allows to access current version of the runner.
 *
 * <p>The current version is parsed from {@link BuildVersion}, but in development mode it can be
 * overridden by setting `enso.version.override` property. This is used in project-manager tests to
 * override the version of projects created using the runner.
 */
final class CurrentVersion {
  private CurrentVersion() {}

  private static SemVer version = null;

  public static SemVer getVersion() {
    if (version == null) {
      version = computeVersion();
    }
    return version;
  }

  private static SemVer computeVersion() {
    var buildVersion =
        (SemVer)
            SemVer.parse(BuildVersion.ensoVersion())
                .getOrElse(
                    () -> {
                      throw new IllegalStateException(
                          "Fatal error: Enso version included in buildinfo is not a valid "
                              + "semver string, this should never happen.");
                    });
    if (!BuildVersion.isRelease()) {
      var overrideVersionProp = System.getProperty("enso.version.override");
      if (overrideVersionProp == null) {
        return buildVersion;
      }
      var parseRes = SemVer.parse(overrideVersionProp);
      if (parseRes.isSuccess()) {
        return parseRes.get();
      }
    }
    return buildVersion;
  }
}
